### DISCUSSION TITLE
Issues / Project Beta at Repo Level

### DISCUSSION THREAD

**Discussion by jonmill** [community/community#6229] [Projects and Issues] [answered]
Hi - looking at the Issues / Projects beta, this looks to be an org-specific feature and not on the repo level. This will start to get very chaotic and unmaintainable for large enterprises with hundreds of repos. Are there any plans to make the Projects be repo-specific as well in order to help larger orgs not become chaotic with thousands of projects used for repo-specific purposes?

Edit for additional questions:
For a more specific ask - are there any plans to replace the Repo-specific Projects tab with the new Projects? That would be incredibly useful to have Projects at the Org level and the same capabilities at the repo-specific level

**Comment by Katsute** [comment]
At the bottom of <https://github.com/features/issues>

> Our roadmap towards general availability (GA) will evolve, but expect improvements to the core flows and new functionality to be continuously added. Example of future roadmap features include:
>
> 1. Ability to create new projects (beta) at both user and repo levels

**Reply by jonmill** [reply]
Ah apologies, I missed that line. Sorry about that, thank you for the reply!

**Comment by azenMatt** [comment]
Hi @jonmill - we are working on this at the moment, 

The way we are thinking about this is less about duplicating the functionality between orgs and repositories and more about allowing repositories to be a place where you can **curate** a selection of projects that will be useful to users of that repository. 

We found that in the first version of projects, having both organisation projects and repository projects caused a lot of confusion on when to pick which type. 

Coming in December, we will release something that looks a little like: 

- The organisation is still the place to create a project 
- Repositories (and in the future teams) will use the existing projects tab as a place where you can curate a list of projects from across your organisation 


![image](https://user-images.githubusercontent.com/7584089/143058470-b81066d1-8333-485d-954e-8b3246da68dc.png)

>order to help larger orgs not become chaotic with thousands of projects used for repo-specific purposes?

This is then also on our radar, but will be further into the future. I see the solution to this as having a more curated /projects tab at the organisation level - rather than just having a date-order list of _everything_. 

Does this feel like it would help with bringing some order to wrangling a large number of projects?

**Reply by spyroskaravanis** [reply]
Hi @azenMatt is there a way in which we can have an early access on this feature. Our company just bought an enterprise account and we are thinking to replace youtrack. It is nice to have issues on company level for roadmap but we really need to set up scrum at repo level.

**Reply by azenMatt** [reply]
@spyroskaravanis - it isn't ready yet, but is in progress. We will have it out to everyone as soon as possible. Likely mid-Dec so not too long!

**Reply by jonmill** [reply]
@azenMatt - This doesn't quite make sense to me...for a large organization, not every repository is related, yet all of the projects will be organization wide? That seems more confusing than having projects that can be org-scoped or repo-scoped, allowing individuals to determine the right level for their situation. To me, this is going to pollute the organization-scoped projects with tons of different projects that have no relation to each other but **must** be there in order to use the new features. 

As an example - my team's project is fairly self-contained at this point, but we'd love to use the new project features. Now, we'll need to create a project at the Microsoft organization scope just for use in our small project; extrapolate that across the entire Microsoft GitHub organization and you can see how messy and disorganized that will be. 

I'd love to chat more about this to understand more on your side about the reasons behind that decision, but also to provide some context around what my team (and others in company-org) are looking to achieve with the new Projects features. Is that something we could setup?

**Reply by azenMatt** [reply]
@jonmill - Would be great to learn more - i'l reach out

**Comment by jensljungblad** [comment]
@azenMatt I think it makes sense, I was actually wondering if it was gonna be something like that. It's a tricky problem, considering you're also solving the "how to create projects and add issues to it from multiple repos in a sane way" which is something we have wanted for a while, so that's cool that it's being addressed with the new projects. But as you say it then leads to the problem with "repo specific projects", and where they should go.

I think if projects are designed to be able to contain information from many different repos, it makes sense they exist on the organization level primarily. I suppose it *could* get messy in the org view if it's just a bucket listing every single project, but as you said, if it too can be curated, maybe not a problem?

Looking forward to trying it out. Good luck!

**Comment by azenMatt** [comment, marked as answer]
We just released this 😁  

https://github.blog/changelog/2021-12-16-the-new-github-issues-12-16-update/

Check it out and let me know what you think!

**Reply by Katsute** [reply]
TBH I would still much prefer projects to be owned by the repository rather than the organization. Nearly all of my project boards are exclusively for those repositories and having them at the organization level just creates unnecessary noise. I feel like we should have the option to create project tables under the repository rather than the organization.

**Reply by joshuaellis** [reply]
I agree with this, as a design agency, our repos don't cross over as part of wider business plans.

**Reply by aembler** [reply]
Strong agree on this as well. It's not just the noise of projects living at the same level alongside all others – it's the friction of creating issues within those projects. I'm looking at starting a project from scratch, with at least 20-30 issues within the project, but the entire project belonging to a single repository. Am I really to create these issues in the sheet view, click on them, and then select the same single repository for every issue as I convert a line item to an issue?

I even created the org-wide project, and added it to the repository project's tab using the new functionality discussed elsewhere in these threads, but it doesn't seem to default these new issues to belonging to that repository.

**Reply by kingma-sbw** [reply]
It needs a paradigm change. An enterprise is no longer repository centered but project centered. So the 'super' workflow would be:

- Create Projects
- Create Teams
- Assign Teams to projects
- Define one or more repositories for a project
- Define Sprints and Tasks  for Projects which trickle down 
- As issues for repositories. 

Am I correct in my assumption that this is a overall workflow the GitHub team considers? It would move away from putting repositories at the center of everything but instead use projects as the center of the univers.

**Reply by bobbrow** [reply]
I have no intention of using beta projects because I have to create them in my (huge) organization. But when I click on the "Projects" tab, GitHub highlights the empty "Projects (beta)" label every time instead of the more useful "Projects (classic)" label which results in having to do an extra click every time I navigate to Projects.  Can you just fold them all together into a single view instead of separating them?  I don't think the distinction of project type requires an actual separation in the UI.

**Comment by LRathje** [comment]
A question: If you create an organizational beta project and curate it at the repository level, can you then ensure users only for this repository also only can see this project? And not all other projects etc.
Another way to ask: Can the new beta project be set up to a structure where you allow co-operationg users per repository?

**Reply by azenMatt** [reply]
Hi @LRathje - not directly. 

Curating it at the repository level has no impact on the permissions. 

The best way to do this currently would be to add individual permissions to the project directly (which can be done via teams - which should help make this easier if you are also using teams to manage repo permissions).

**Reply by LRathje** [reply]
Okay thanks. So if I setup the project, invite people, then they will be limited to this project. And if I only curate to a single repository they will not see other repositories. But then, I guess, they should also been given access to the repository to see everything there?
Lars

**Reply by azenMatt** [reply]
Yes exactly - and any item which is from a repository that a user does not have access to will be shown as:

![image](https://user-images.githubusercontent.com/7584089/149950080-71a4cf0c-fe21-4f3d-a014-d23760985c6c.png)

So really, a user would need access to both the project and the repos added to it. But if you do need to mix in some more restricted items into a project, you can still do that and not worry about anyone seeing something they shouldn't

**Reply by LRathje** [reply]
Great - thanks

**Comment by electrovir** [comment]
The following scenario related to repo-level projects is keeping me away from Projects Beta and leaving me very worried about what to do when this beta stamps out the original projects tab:

1. GitHub user A creates repo Q within the scope of their own profile.
2. User A adds collaborator GitHub users B and C to repo Q.
3. With Projects Beta, GitHub user A is now required to micromanage creating projects for repo Q.

If user A is taking a leadership role and not actually actively working on repo Q, it's not feasible for them to manage projects for repo Q. It also pollutes user A's user project space with items irrelevant to their actual work.

It also doesn't make sense for these three users to form a whole new GitHub organization if they're only collaborating once probably ever in their whole lives for repo Q. (A good example of this would be a small one-off group project in school.)

**Reply by LRathje** [reply]
Interesting scenario. Waiting for what is the answer or suggestions from @azenMatt

**Reply by azenMatt** [reply]
Hi - that's an interesting scenario, if I can rephrase - your concern is that: 

> In the old version of projects, a user who had been added as a collaborator to a repo could create a new project as part of that repo. In Projects beta that is not possible as the projects live at the organization/user level they must be created by an org-member / user 

A couple of thoughts here: 

- With the free teams plan, creating a one-off organisation could be a solution. (then, for example with the school project - all users could have the same level of access) 
- If one of the users wants to keep everything under their username, then they would need to make sure the others are added as collaborators. 

Let me know if you find this comes up frequently in day-to-day. But it does sound like this user would benefit from being on a team plan.

**Reply by TechWatching** [reply]
Same concern here. I have created a repository under my profile. It's just one repository, I don't need an organization for that. I have invited 2 friends to contribute to the repository. As I can't create a project at the repository level I have to create it on my profile as a private project and invite my 2 friends as collaborators of this project. The project is to organize our work on this repository but nothing links it to the repository.

**Comment by LRathje** [comment]
Now that I better understand how the ownership and user management works between org-projects (beta) and repositories, I'm still wondering how to organize the expected mess of many, many projects at the org level, that maybe for more than 90% each live solely within a single repository?

**Comment by urapadmin** [comment]
If we were given a choice whether to manage our projects on repo or on org level, we would go for repo level. We use projects a lot to bundle issues by topic and workflow on repository level and having all those projects turn up on organizational level once the beta phase is finished will be a mess. 

Also: I was hoping that UN-checking the "Enable projects for the organization" level would do exactly that. But instead one is left with no way of creating a beta project anymore anywhere. Personally I am still confused by the new concept.

**Comment by IanMitchell** [comment]
I also want to chime in that as a tester I'm confused by the new model and prefer v1 repo-based projects. I don't use organizations in my workflow heavily, and it feels like this is pushing very very hard in that direction - I think I'm more likely to not use projects than to adopt organizations, but as someone heavily using v1 projects today I would be massively bummed out about that.

One question I have which I don't want to test manually - what happens when I transfer repository ownership to someone else, but I own all the projects? Can I hand those over too, or does that user have to recreate them entirely?

**Comment by bestickley** [comment]
Maybe another improvement towards projects feeling more like single repo is to allow setting the default repository per project so you don't have to select it each time converting draft issue to real issue?

**Comment by staplespeter** [comment]
Hello, new to Github and was starting out a new project.  I'll add my thoughts as a noob, though please let me know if I am using the features incorrectly or have missed features I should be using.  I am using a User project and not an Organisation project.

I want to have a project that will consist of multiple systems and thus repos, with a list of project related tasks and documentation.  Ideally i would like the documentation stored at the User project level. There are no features for this.  I created a repo for the documentation and linked the document related tasks to the repo as issues.  This is an acceptable solution that works for the old, and beta, User level projects.

One of the User project tasks is to generate a list of user driven features (e.g. Add login feature).  I'd like them to be visible as tasks at the User project level.  Ideally, at the User project level, I would be able to create a sub-list of tasks for these tasks, splitting the user driven feature into a multitude of other tasks, each representing a repo feature, and thus linked to a repo.  This is just an organisational method so I'm happy enough to split a user driven feature (that would result in changes to several different repos) into several similarly named tasks (system features), which are linked to the individual repos (e.g. Add login feature -> Add login front-end, Add login back-end).  This works for the old, and beta, User level projects.

For each of these system features, using the old repo based projects, I can then manage it within the individual repo by creating a new project for that system feature and creating a list of tasks there.  Unfortunately the beta Project does not support this, as there are only User level projects in the beta, and only these can be linked to the Repo as it's projects.  This means that the beta User project (as others have mentioned) will become crazy polluted with what should be repo level projects, making it unmanageable.

I know this isn't a complex project management application but that kind of hierarchical organisation of tasks is a simple feature of management and essential in my eyes.  I do like that tasks can be added in spreadsheet from, so the new interface in the beta is definitely preferable.

**Reply by LRathje** [reply]
Also this scenario calls for projects on both levels. You and so many other github users are requesting this clearly.
So why can't the beta project features not just be extended to the repository level, please!!!

**Comment by lyndseypadget** [comment]
I appreciate the progress and transparency on GitHub projects overall.  However, I am also unable to get on board with Projects Beta due to this issue.  While the ability to link many repos to a project is a "nice to have", the ability to control the visibility/access of the project is a "must have".

My scenario is:

1. My company is a GitHub organization.  Within it, there are teams.  I have several repos for my project, all of which are private.  I grant users access to these repos by making them part of a GitHub team.  Of course, that team is part of the organization.
2. Our current project board (not Beta, kanban style) belongs to one of the repos, where all of the issues reside.  This is considered the "primary repo".  Because the repo is private and only accessible to the team(s) given access, the board is also private.  This is important.
3. I created a new project (Beta, spreadsheet style) at the org level.  The project's visibility is private, and there does not appear to be a base role that aligns with the level of access I need to give to team members.
4. I asked a member of the organization _who is not a member of the team_ to go to https://github.com/orgs/OURORGNAME/projects?type=beta.  She could see the project, which is unacceptable.

Am I missing something?

I don't necessarily feel strongly that the project should be at the repository level.  In fact, I agree with @kingma-sbw that projects are, well, projects... and this concept does not yet exist in GitHub.  In lieu of that, projects at the repository level is good enough.

What I do feel strongly about is that people who aren't part of the team shouldn't be able to see the project.

**Reply by azenMatt** [reply]
Hi @lyndseypadget - the recommend action to take here would be to: 

1. Change the base role to no access [1]
2. Add your team directly to the project [2]

Does this help solve the access problem for you?

[1]  ![image](https://user-images.githubusercontent.com/7584089/164454381-c728f128-3872-4624-a967-1ac6aafaefa7.png)
[2] ![image](https://user-images.githubusercontent.com/7584089/164454570-8de91da1-e353-4836-bb76-26f2bd3fd8cc.png)

**Reply by lyndseypadget** [reply]
Thanks for the direction @azenMatt.  The answer is: yes and no.  Fair warning, the following is going to sound pedantic, but hear me out.

These instructions do restrict or grant access as one would expect (whew).  But, the language under "No access" is misleading.  The first sentence implies that the project being public is a _prerequisite_ for the action in the second sentence.  This seemed too visible so I discounted it as a viable choice.

- see = visible to
- access = permissions to
- additional = in addition to the thing that was just stated (the project being public)

The project's visibility is straightforward: Private means only permitted people can see it, and public means everyone in the world can see it.  Roles / permissions are muddier.  In my view, project access should be managed in the same way as existing GitHub Projects, via these two avenues:

- the team at https://github.com/orgs/OURORGNAME/teams/OURTEAMNAME/projects
  - This does not allow me to add a beta Project today.  If and when this happens, can the "Base role" section go away or take the form of classic Project permissions?
- the project at https://github.com/orgs/OURORGNAME/projects/PROJECTID/settings
  - I noticed that the language under "None" on classic Projects is the nearly the same 🤦‍♀️🤷‍♀️

This might be somewhat tangential to the core issue, but seems closely related.  Happy to log new issue(s) if they seem valid but out of scope.

**Comment by DanielNoord** [comment]
@azenMatt I think the sentiment in many of the answers after this was marked as "answered" is the same (see [for example this comment](https://github.com/github-community/community/discussions/6229#discussioncomment-3004723)). Perhaps this was not the original intent of this question, but many end up here wondering if there will ever be an option to create the new type of projects on the repo-level (it's also the first Google result for searches for `GitHub projects beta repository level`). You can scan this discussion for various arguments as to why projects on the organisation level are not always practical or possible. I can also give you some if there need to be more use cases for this to be put on a milestone.

I have been keeping an eye on this discussion and the various roadmaps and milestones but still haven't seen a clear and definitive answer to the main question:
Is Github considering at some point allowing the creation of projects at the repository level? Or is this something that will never be possible?

If not, I (and I think many others) can stop following this discussion and just keep using the old projects. Right now I keep hesitating to create new projects as I hope to one day be able to use repo-level beta projects

**Reply by azenMatt** [reply]
Hi @DanielNoord 

We are still working on improving how this works. 

>Is Github considering at some point allowing the creation of projects at the repository level? Or is this something that will never be possible?

Impossible would be the wrong word to use, but our hope is to only have one 'type' of project to avoid the concept of 'repository' project and 'organization' project. We found in research of the old projects this confused a lot of people. (although I acknowledge that it's absence also causes confusion!) 


There are two things we are looking at to improve how this works, one of which we just released, the other which will come later. 

- **Recent projects** This just shipped. If you go to the /projects page on an org, you will notice that the default view is now for projects you have recently accessed. 

This should both be a nicer experience and also address the concerns I have been hearing that repository projects are nicer as they don't create noise at the org level. As a solution, these recent projects are designed to not make the projects landing page at the org feel 'messy' or full or projects that you don't care about. 


- **Repository permissions** This is a bigger task that we have not started yet. Right now, a project can have permissions based on either teams or users. What if a project could have a permission which references a repository? 

If we could do this, then we could let you create a project from a repository - when that project is created, it is still an 'org' project - but its permissions are mapped to the repository where it was created. (and of course - you would be able to change this if your needs change later) 

This means: 
- You don't have several places you need to manage permissions. If you have access to the repo, you will have access to this project. 
- At the org project page - only users who have these permissions will see it. Creating a cleaner space at the org level while still being the single source or truth where you can see all your projects.


What do you think, would either of these help with your concerns?

**Reply by leandro-lucarella-frequenz** [reply]
I have a use case that seems to be hard to overcome without repository-specific projects. We use a project per release for a particular repository (as a side note, we used to use milestones for this, which make more sense conceptually, but milestones are so far behind in terms of features, that is not practical anymore). These projects are really only relevant to the repository and if they are hanging from the organization, at a bare minimum we should name them "[repository] [version]" instead of just "[version]". And again, if you have that times 50 or a 100 repositories with their own release projects, that is a lot of projects, is hard to understand why anyone would want to have them all related to the organization instead of individual repositories.

It is still hard to understand why the concept of a per-repository projects can be so hard to understand to users when all the other per-repository tabs are fine (issues, wiki, discussions, etc.).

**Reply by DanielNoord** [reply]
@azenMatt Thanks for the detailed response. 😄  I left some (personal) considerations.

> Impossible would be the wrong word to use, but our hope is to only have one 'type' of project to avoid the concept of 'repository' project and 'organization' project. We found in research of the old projects this confused a lot of people. (although I acknowledge that it's absence also causes confusion!)

If it were up to me I would reconsider this. This issues raised in this discussion show that this choice might have caused unexpected additional confusion. In this case I'm not sure the solution for the previous confusion is to remove the distinction. Why not try and make those distinctions clearer with visual guides and hints? The current project page already has two tabs (old and new), can those be organisation and repository? Just to name a suggestion.

> If we could do this, then we could let you create a project from a repository - when that project is created, it is still an 'org' project - but its permissions are mapped to the repository where it was created. (and of course - you would be able to change this if your needs change later)

This seems like it should solve the current issues. 👍 But... 

> * You don't have several places you need to manage permissions. If you have access to the repo, you will have access to this project.

I can understand this might cause confusion, but this seems to narrow or simplify real world cases. (Open-source) projects organisational hierarchies often aren't easily reducible to a set of different permissions levels within one organisation. The additional levels that repositories give can be very useful.

> * At the org project page - only users who have these permissions will see it. Creating a cleaner space at the org level while still being the single source or truth where you can see all your projects.

This requires a paradigm shift for my personal workflow and I guess for many others. I work from repositories and occasionally deal with organisations. I almost never visit organisational pages as I don't really use those. For me the repository project page should be "clean" and the single source of truth.

> It is still hard to understand why the concept of a per-repository projects can be so hard to understand to users when all the other per-repository tabs are fine (issues, wiki, discussions, etc.).

I think this, although phrased a bit strongly, sums up perfectly what's the issue for many here. Perhaps we weren't vocal enough in the previous research but (apparently) many people have a repository-centred workflow rather than a organisation-centred workflow. The current proposal seems to create some imaginary organisation which would then allow the creation of a repository scoped project. Creating imaginary organisations will likely only increase confusion over what controls what. 

As I said above, I would recommend to try and solve the confusion identified by the previous research by being more explicit rather than trying to simplify the real-world and narrow down the possibilities for users.

**Reply by azenMatt** [reply]
>For me the repository project page should be "clean" and the single source of truth.

What we are proposing here should help - as we could list any project which has this repository as an explicit permission in the repository (as well as the organization). We already started down this path with the ability to 'add' a project to a repository as a quick link) 

That way - you could effectively never have to go to the organisation page. Everything you want to do with a project can be done from the repo (Create, access, manage)

**Reply by DanielNoord** [reply]
That does indeed sound like a step in the right direction for my personal workflow.

Still I would ask you to reconsider the proposal of creating repository projects from (imaginary) organisations. 

Perhaps I’ll be one of those who expresses confusion in a future survey, so if you go down this route at least be very clear what the purpose of such organisations would be and why I would want to create one.
I’m not sure I’ll be able to convince many of the maintainers of projects I contribute to to create an organisation and pay the associated additional permissions churn/maintenance just for the sake of using the projects beta.

**Reply by Katsute** [reply]
I would rather have project tables be _literally owned_ by the repository, rather than organizational projects _pretending_ to be a repository project.

**Recent projects** still have _repository specific_ projects mixed in with organizational projects. I would rather have organizational projects have **only organizational projects**.

Currently project tables are _physically detached_ from the repository, if I open a classic project I can access all the resources in the repository (code, issues, pulls, discussions), if I open a project table it redirects to the organization.

Having organization projects _pretending_ to be a repository level project simply isn't the same as having a project living under a repository.

**Reply by justlevine** [reply]
@azenMatt perhaps I just don't understand it, but would the above solve the use case for community driven development mentioned in earlier comments?

I've seen and been part of many repos where the owner is little more than stewards, and active development is carried on by repo collaborators, who use repo-level projects to coordinate development. The dependence on an organization-level approval can be a huge bottleneck, and the cross contamination of organizational data with repo data (and real-world workflows) adds to the noise and hinders productivity. 

(As an aside, teams are also not always a good option for organizations, as they can imply direct affiliation of an individual to an organization, and I've interacted with numerous orgs who were hesitant to add a non-employee to an organization's team).

**Comment by Katsute** [comment]
I do not appreciate how GitHub is **forcefully** pushing project tables onto us without adding this one critical feature.

I can't even create classic projects for new repositories or even use the API to create a project.

[![docs](https://user-images.githubusercontent.com/58778985/186807896-0e8d106a-7163-4024-b24b-0f933003a3bc.png)](https://docs.github.com/en/issues/organizing-your-work-with-project-boards/managing-project-boards/creating-a-project-board#creating-a-repository-classic-project)

[![api](https://user-images.githubusercontent.com/58778985/186806539-27427e49-73ee-4047-b69d-119c6957770d.png)](https://docs.github.com/en/rest/projects/projects#create-a-repository-project)

If repository level projects **are not even being considered**, stop preventing us from using classic projects. I do not operate like an organization and do not want these projects stored at the organizational level.
